cosmology/StructureFormationManager.h

```diff
@@ -0,0 +1,597 @@
+            mes << "Starting first repartition" << endl;
+            this->isFirstRepartition_m           = true;
+            this->loadbalancer_m->initializeORB(FL, mesh);
+            this->loadbalancer_m->repartition(FL, mesh, this->isFirstRepartition_m);
+        }
+
+        static IpplTimings::TimerRef ReadingTimer = IpplTimings::getTimer("readData");
+        IpplTimings::startTimer(ReadingTimer);
+
+        std::ifstream file(this->folder + "Data.csv");
+
+        // Check if the file is opened successfully
+        if (!file.is_open()) {
+            std::cerr << "Error opening IC file!" << std::endl;
+        }
+        
+        // Vector to store data read from the CSV file
+        std::vector<std::vector<double>> ParticlePositions;
+        std::vector<std::vector<double>> ParticleVelocities;
+        double MaxPos;
+        double MinPos;
+
+        // Read the file line by line
+        std::string line;
+        int i = 0;
+        while (std::getline(file, line)) {
+            std::stringstream ss(line);
+            if(i % ippl::Comm->size() == ippl::Comm->rank()){
+
+                // Read each comma-separated value into the row vector
+                std::string cell;
+                int j = 0;
+                std::vector<double> PosRow;
+                std::vector<double> VelRow;
+                while (j < 6 && std::getline(ss, cell, ',')) {
+                    if (j < 3){
+                        double Pos = std::stod(cell);
+                        PosRow.push_back(Pos);
+                        ++j;
+                        // Find Boundaries (x, y, z)
+                        if(i+j > ippl::Comm->rank()){
+                            MaxPos = std::max(Pos, MaxPos);
+                            MinPos = std::min(Pos, MinPos);
+                        }
+                        else{ // very first input
+                            MaxPos = Pos;
+                            MinPos = Pos;
+                        }
+                    }
+                    else {
+                        double Vel = std::stod(cell);
+                        VelRow.push_back(Vel);
+                        ++j;
+                    }
+                }
+                ParticlePositions.push_back(PosRow);
+                ParticleVelocities.push_back(VelRow);
+            }
+            ++i;
+        }
+
+        // Boundaries of Particle Positions
+        mes << "Minimum Position: " << MinPos << endl;
+        mes << "Maximum Position: " << MaxPos << endl;
+        mes << "Defined maximum:  " << this->rmax_m << endl;
+
+        // Number of Particles 
+        if (nloc != ParticlePositions.size()){
+            std::cerr << "Error: Simulation number of particles does not match input!" << std::endl;
+            std::cerr << "Input N = " << ParticlePositions.size() << ", Local N = " << nloc << std::endl;
+        }    
+        else 
+            mes << "successfully done." << endl;
+
+
```

Comment:
add comment here explaining that the particle positions and velocities which are read in above form the initial conditions file are assigned to the particle attributes R and V in the particle container